### PR TITLE
🎛️ Replace base `ExpandingFormset` functionality with Stimulus FormsetController `w-formset` inc. initial adoption

### PR DESCRIPTION
**Pull request by lb-**
First stage of #7915

## Overview

* Introduces a new Stimulus class `FormsetController` with the identifier `w-formset` with the longer term plan for this to be used for the InlinePanel JS functionality also.
* Introduces a refactor of our common code for Django's formsets that streamlines how we set the delete input widget and prepares some of the Stimulus data attributes for use in the templates. This is a new base mixin `wagtail.admin.forms.formsets.BaseFormSetMixin` (happy for a different name to be suggested).
* Replaces only the basic expanding formset usage in Group edit/create views, Workflow edit/create (Page selection only) with the Stimulus usage. No changes to `InlinePanel` at all.
* Most importantly, this new approach is CSP compliant, has validation error messaging feedback for developers and does not rely on any complex id string wrangling, with fully tested support for complex nested usage.
  * Stimulus does most of this for us through the [Stimulus Scopes](https://stimulus.hotwired.dev/reference/controllers#scopes) feature.
  * I've left all nested IDs in place in the templates as there is no major need to remove these for now, but they're no longer used for any JS behaviour.
  * Relates to getting closer to CSP compliance #7053 & #1288

### Additional notes

* Introduces event dispatching for the expanding formsets to align with the `InlinePanel` https://docs.wagtail.org/en/stable/reference/pages/panels.html#inline-panel-events (have not documented, I thought release notes would be enough for this as it's not really used by customised code).
* Removes the reliance on jQuery animations by introducing a function (which we can abstract later if it's useful) that uses a promise callback for 'hiding' after an animation/transition is completed on an element without having to hard-code how long we expect the animation to take. For example, we can change the `selected` class to be ...-200 and the hidden attribute will be set after 200ms instead of the current 300. This reduces a common issue where we get the animation timing and the attribute updating out of sync causing chunky animations or confusing delays.
* Commits are broken up to be a step by step creation and adoption of this new code, hopefully making reviewing easier.
* instead of style display none, now using `hidden` for when a child form is deleted. This is more appropriate for the UI as the element is essentially meant to be removed. This also makes CSP issues with style attribute changes go away.
* There will be a subtle change in behaviour between the current, if you add 1001 children in any of the usage below. Currently, this would fail on server submit, however, with this change the add button will simply not add any more. I can change this to introduce a new option to ignore min/max but it seemed to make sense to keep this initial adoption simpler even with that edge case behaviour difference.
* Naming of 'formset' vs 'form_set'
  * I have opted for `FormSet` in the Python class names, this is what Django uses.
  * `formsets.py` for the file in our admin files, again, following the Django convention.
  * `FormsetController`/`w-formset` (not `FormSetController`/`w-form-set`) in the Javascript code for brevity and alignment with our existing set up event names on the InlinePanel code (e.g. `w-formset:added`).
* Naming of the `'removed'` event vs (`delete`/`deleted` action/code).
  * Django uses the convention `removed` for the event https://docs.djangoproject.com/en/5.0/ref/contrib/admin/javascript/ we use that same event name in our documentation for InlinePanel https://docs.wagtail.org/en/stable/reference/pages/panels.html#javascript-dom-events
  * Django uses the names `delete`/`deleted` in the code.
  * I have aligned with that approach in the formset, while it's a bit odd to have `removed`/`removing` in the event names I hope this clarifies why.
* Additional clean up benefits include reduction of duplicate code that was imported across views, better leverage of Python / Django mixins to support potential Python side customisation behaviour in the future, and increased Python test coverage over the data attribute generation.

## Checklist

-   [X] Do the tests still pass?
-   [X] Does the code comply with the style guide?
-   [X] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments? Chrome, Safari, Firefox all manually tested
-   [X] For new features: Has the documentation been updated accordingly? (Upgrade considerations added)

## How to test

1. Workflow edit, workflow create - validate that the pages sections all update correctly when adding/removing.
2. Workflow edit for a disabled workflow, check the console has no errors and the content is not visible for page selection as per current code.
3. User group create/edit - validate the page permissions and other non-page permissions changes.
4. For the above, check things still work as expected on Save with a POST form failure.

### Deprecation of `window.buildExpandingFormset` global function

The global function `window.buildExpandingFormset` to attach basic JavaScript behavior for Django's expanding formsets has been deprecated and will be removed in a future release.

Using `InlinePanel` will not be impacted, this only impacts a small set of basic expanding formsets in use across Workflow and Group view editing.

Previously these expanding formsets required a mix of specific id attribute structures and inline scripts to instantiate with callbacks for handling deletion.

In the future, the following data attributes are required to emulate the same behavior, these are likely to change and should not be considered official documentation.

| Element                                      | Attribute(s)                              |
| -------------------------------------------- | ----------------------------------------- |
| Containing element                           | `data-controller="w-formset"`             |
| Element to append new child forms            | `data-w-formset-target="forms"`           |
| Child form element                           | `data-w-formset-target="child"`           |
| Deleted form element                         | `data-w-formset-target="deleted" hidden`  |
| `template` element for blank form            | `data-w-formset-target="template"`        |
| Management field (total forms)               | `data-w-formset-target="totalFormsInput"` |
| Management field (min forms)                 | `data-w-formset-target="minFormsInput"`   |
| Management field (max forms)                 | `data-w-formset-target="maxFormsInput"`   |
| Management field (Delete, within child form) | `data-w-formset-target="deleteInput"`     |
| Add child `button`                           | `data-action="w-formset#add"`             |
| Delete child `button` (within child form)    | `data-action="w-formset#delete"`          |

Usage of nested id structures are no longer required but can be left in place for easier debugging.

### CODE REVIEW

File: client/src/controllers/FormsetController.ts [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+      target.contains(input),
+    );
+
+    if (!deleteInput) {
+      throw new Error(
+        `Could not find "deleteInput" target within removed form. ${target.nodeName} with id '${target.id}'.`,
+      );
+    }
+
+    deleteInput.value = '1';
```
Comment (lb-) [inline client/src/controllers/FormsetController.ts:235]: I should probably not set the value if already set to this and hence not dispatch a change event unless changed. 

Solves for an edge case on POST when the form has deleted items already in the response HTML.

File: client/src/controllers/FormsetController.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    });
+  }
+
+  /**
+   * When removed, add the class and update the total count.
+   * Also update the DELETE input for this form.
+   */
+  childTargetDisconnected(target: HTMLElement) {
+    const deletedClasses = this.deletedClasses;
+
```
Comment (lb-) [inline client/src/controllers/FormsetController.ts:215]: We can probably check if the element is already hidden and bypass this logic. Might be a better and simpler approach to ensure no events get dispatched on POST error scenarios when the deleted items may already be in the response.

File: client/src/controllers/FormsetController.ts (hunk 1 above)
Comment (lb-) [reply client/src/controllers/FormsetController.ts:237]: Fixed & rebased.

File: client/src/controllers/FormsetController.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    )(target.classList.add(...deletedClasses)).then(() => {
+      target.setAttribute('hidden', '');
+      this.dispatch('removed', { target, cancelable: false });
+    });
+
+    const deleteInput = this.deleteInputTargets.find((input) =>
+      target.contains(input),
+    );
+
+    if (!deleteInput) {
```
Comment (lb-) [inline client/src/controllers/FormsetController.ts:229]: Note. I could not easily write a unit test for this error case due to Stimulus not catching these errors. 

https://github.com/hotwired/stimulus/issues/711

I have manually tested this scenario though and it does correctly throw in the console without breaking the whole Controller behaviour.

File: wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html [hunk 4]
```diff
@@ -45,30 +45,14 @@
...
-        <tr id="inline_child_{{ formset.empty_form.prefix }}">
+    <template id="id_{{ formset.prefix }}-EMPTY_FORM_TEMPLATE" data-w-formset-target="template">
+        <tr id="inline_child_{{ formset.empty_form.prefix }}" data-w-formset-target="child">
             {% include "wagtailadmin/permissions/includes/collection_member_permissions_form.html" with form=formset.empty_form only %}
         </tr>
     </template>
 
     <p class="add">
-        <button class="button bicolor button--icon" id="id_{{ formset.prefix }}-ADD" value="Add" type="button">{% icon name="plus" wrapped=1 %}{% block add_button_label %}{% endblock %}</button>
+        <button class="button bicolor button--icon" id="id_{{ formset.prefix }}-ADD" type="button" data-action="w-formset#add">{% icon name="plus" wrapped=1 %}{% block add_button_label %}{% endblock %}</button>
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html:55]:
```suggestion
        <button class="button bicolor button--icon" id="id_{{ formset.prefix }}-ADD" value="Add"  type="button" data-action="w-formset#add">{% icon name="plus" wrapped=1 %}{% block add_button_label %}{% endblock %}</button>
```

Accidentally removed `value="Add"`. I actually think we should not be setting the value on these buttons as it's not used visually and I'm pretty sure it's not used by the server.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#value

Probably an investigation for a different issue.

**Comment by lb-** [comment]
While this PR is not a fix for https://github.com/wagtail/wagtail/issues/11760 it's a good note that developers are expecting all features to work across ModelViewSet / Snippet views without having to think about what JS to import.

Hence, I am keen to get on the task of migrating `InlinePanel` once this PR is in.

**Comment by lb-** [comment]
Note to self. Rebase time.

**Comment by lb-** [comment]
Rebased plus found a problem with scripts within templates which I have fixed and included unit tests for.

I have also refined some of the deprecated notes & JSDoc to be a bit clearer.

**Comment by lb-** [comment]
OK - this may not be the direction we will be going in for InlinePanel / Expanding FormSets (see https://github.com/wagtail/telepath-forms ) - will leave with you @gasman 

Happy if you feel I should close this PR or move to draft, for now I will mark as needs design decision.

**Comment by thibaudcolas** [comment]
Assigning to you two to either make this design decision, or otherwise take it to a bigger group to outline your thought process.

**Comment by lb-** [comment]
Had a discussion at core team, @gasman will be preparing a plan for the core team on what a different migration path could look like for formsets, modal workflows and related modules.

Thanks @thibaudcolas

**Comment by lb-** [comment]
I am still of the strong opinion that this is the best path forward for our codebase to remove jQuery, avoid inline scripts and to rely less on fragile 'id' matching for nested formsets. I really think this gives us a great path to build on this to solve the other core jQuery based dynamic formsets and even parts of the modal workflow.

Please take another look at this @gasman - it would be great to get some clarity in this space if possible. I am not trying to cause friction, hopefully this comes across OK. I know there's other chats in the core team slack also.

in light of a different plan and direction from that laid out in the RFC https://github.com/wagtail/rfcs/pull/78 - I have opted to rebase this & fix conflicts. I have also moved the upgrade considerations to the PR description to avoid continued conflicts with this PR and `main`.

Fun times ahead. Let's see what happens :D

**Comment by lb-** [comment]
Just rebased plus pulled out the `transition` promise utility to it's own file and also re-used the util `runInlineScripts` from @gasman  in #11839 (I had forgotten this was added, super handy).

I think longer term we may add an option like 'unsafeRunInlineScriptsValue` to the `FormsetController` but I think we are still a while off from being able to do that ourselves within the admin. But, it's nice to use this shared function to track this for the future.

**Comment by lb-** [comment]
CI issue is the semgrep issue, not related to this PR.
> FileNotFoundError: [Errno 2] No such file or directory: '/home/circleci/project/.venv/bin/semgrep'

**Review by gasman — APPROVED** [review]
It's taken a while for me to get my head around it all, but I'm very happy with all of this - a great improvement!

I especially like the use of `deletion_widget` to eliminate the messy patching of the delete controls on individual forms - great job spotting that - and the fact that we effectively get nesting support for free thanks to [Stimulus's controller scoping rules](https://stimulus.hotwired.dev/reference/controllers#nested-scopes).

**Comment by gasman** [comment]
Merged in 9889bddcb8823369f7a26c6b6f4a3eb6784d5985

**Comment by lb-** [comment]
Thanks @gasman - I appreciate the review and also the discussions in core team, it helped me take a few extra looks at this change and find issues or confusing aspects. 

Hopefully this creates a great building block for the rest of inline panels. 

I'll look at the next PR in the new year if no one else gets to it before me.

**Pull request closed without merging**